### PR TITLE
 STM32 LPTIM driver doesn't restart counter after sleeping K_TICKS_FOREVER

### LINKED ISSUE #25952: STM32 LPTIM driver doesn't restart counter after sleeping K_TICKS_FOREVER
**Describe the bug**
The `z_clock_set_timeout()` function of the STM32 LPTIM driver checks if it is being passed `K_TICKS_FOREVER` and in that case just resets and stops the LPTIM. However passing such a value doesn't mean the system has to halt forever. The system might just wait for an interrupt for an undetermined amount of time, for example the end of a I2C or SPI transfer. It means that the next time `z_clock_set_timeout()` is called, the LPTIM should be re-enabled and reconfigured. Otherwise the system just hang.

Note that it is a minor issue as the scheduler only passes `K_TICKS_FOREVER` in case `CONFIG_SYSTEM_CLOCK_SLOPPY_IDLE=y` (which is not the default), otherwise INT_MAX is passed instead.

**To Reproduce**
Build `samples/basic/blinky` on a board supporting LPTIM with the following options:
```
CONFIG_DEVICE_POWER_MANAGEMENT=y
CONFIG_STM32_LPTIM_TIMER=y
CONFIG_SYSTEM_CLOCK_SLOPPY_IDLE=n
```
Check that the LED blinks correctly.

Then set `CONFIG_SYSTEM_CLOCK_SLOPPY_IDLE=y`. The LED doesn't blink anymore as the system just hangs.

### PR DESCRIPTION
**Pull request by FRASTM**
This patch changes the LPTIM when setting a timeout 'FOREVER' 
In this case, the LPTIM clock is stops and the systems goes to sleep until wakeUp. Because it is stopped, the LPTIM can no more exit the sleep mode but another wakeup authorized source could.
Later on, when LPTIM has to be re-enabled, just restore its clock with `LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_LPTIM1);`
This is for example the user button like in the sample application ; samples/subsys/power/stm32_pm/

This fixes https://github.com/zephyrproject-rtos/zephyr/issues/25952

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

File: drivers/timer/stm32_lptim_timer.c [hunk 1]
```diff
@@ -185,14 +185,8 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 		return;
 	}
 
-	/* ARROK bit validates previous write operation to ARR register */
-	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
-	}
-	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
-
```
Comment (erwango) [inline drivers/timer/stm32_lptim_timer.c:192]: This part of the change is not explained

**Review by erwango — CHANGES_REQUESTED** [review]
Please remove the 3 first commits as they don't seem to be needed for the change

**Comment by FRASTM** [comment]
Update after the merge of the  #25940 
Removing the sample application which demonstrate the exit from a sleep mode 'FOREVER'

File: drivers/timer/stm32_lptim_timer.c (hunk 1 above)
Comment (FRASTM) [reply drivers/timer/stm32_lptim_timer.c]: N/A  anymore, now merged in the #25940

**Review by erwango — APPROVED** [review]

**Review by aurel32 — COMMENTED** [review]
I just gave it a try and it doesn't fix the issue.

The fact that the LPTIM peripheral is not reset is a first step. However it is still stopped, and never restored. I guess a flag need to be added to remember has been stopped and in that case, re-enable it on the next call of `z_clock_set_timeout` without `K_TICKS_FOREVER`.

I have just tried to re-enable it unconditionally just after the block checking for `K_TICKS_FOREVER` and it indeed fixes the issue.

**Comment by FRASTM** [comment]
You mean in the set_timeout() :
```	if (ticks == K_TICKS_FOREVER) {
		/* disable LPTIM clock to avoid counting */
	} 
       else if ( LPTIM clock is not enabled) {
		/* enable LPTIM clock to continue counting */
	}
```

I did it on the power exit state, but in case the PM is not enabled this is missing.
I am adding this.

**Comment by aurel32** [comment]
> Do You mean in the set_timeout() :
> 
> ```
> if (ticks == K_TICKS_FOREVER) {
> 		/* disable LPTIM clock to avoid counting */
> 	} 
>        else if ( LPTIM clock is not enabled) {
> 		/* enable LPTIM clock to continue counting */
> 	}
> ```
Yes, this is what I have in mind. Beside the flag test it's what I have done so it should just work.

**Comment by FRASTM** [comment]
Thanks @aurel32

Need to restart lptim clock in the set_timeout if disabled by a previous set_timeout('FOREVER')

File: drivers/timer/stm32_lptim_timer.c [hunk 2]
```diff
@@ -178,12 +178,16 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
...
-		/* disable LPTIM */
-		LL_APB1_GRP1_ForceReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+		/* disable LPTIM clock to avoid counting */
 		LL_APB1_GRP1_DisableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
 		return;
 	}
 
+	/* if LPTIM clock was previously stopped, it must now be restored */
+	if (!LL_APB1_GRP1_IsEnabledClock(LL_APB1_GRP1_PERIPH_LPTIM1)) {
+			LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
```
Comment (aurel32) [inline drivers/timer/stm32_lptim_timer.c:188]: There is an indentation issue here.

**Review by aurel32 — CHANGES_REQUESTED** [review]
Thanks for this new version, I confirm it fixes the issue. I had it mind to keep the state in a variable, but it's indeed much better to just query the state as you did.

Overall that looks good, there is a small indentation issue to fix. Please also mention that it fixes
#25952 in the commit message.

File: drivers/timer/stm32_lptim_timer.c (hunk 2 above)
Comment (FRASTM) [reply drivers/timer/stm32_lptim_timer.c]: changed

**Review by aurel32 — APPROVED** [review]
Thanks, LGTM.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**